packages/beacon-node/src/monitoring/system.ts

```diff
@@ -5,264 +5,211 @@ import {Logger} from "@lodestar/utils";
 
 type MiscOs = "lin" | "win" | "mac" | "unk";
 
-// static data only needs to be collected once
-let staticDataCollected = false;
-// disk I/O is not measurable in some environments
-let diskIOMeasurable = true;
-
-let cpuCores = 0;
-let cpuThreads = 0;
-let cpuNodeSystemSecondsTotal = 0;
-let cpuNodeUserSecondsTotal = 0;
-let cpuNodeIdleSecondsTotal = 0;
-let memoryNodeBytesTotal = 0;
-let memoryNodeBytesFree = 0;
-let memoryNodeBytesCached = 0;
-let memoryNodeBytesBuffers = 0;
-let diskNodeBytesTotal = 0;
-let diskNodeBytesFree = 0;
-let diskNodeReadsTotal = 0;
-let diskNodeWritesTotal = 0;
-let networkNodeBytesTotalReceive = 0;
-let networkNodeBytesTotalTransmit = 0;
-let miscNodeBootTsSeconds = 0;
-let miscOs: MiscOs = "unk";
-
-/**
- * Collect system data and update cached values
- */
-export async function collectSystemData(logger: Logger): Promise<void> {
-  const debug = (dataType: string, e: Error): void => logger.debug(`Failed to collect ${dataType} data`, {}, e);
-
-  await Promise.all([
-    collectStaticData().catch((e) => debug("static system", e)),
-    collectCpuData().catch((e) => debug("CPU", e)),
-    collectMemoryData().catch((e) => debug("memory", e)),
-    collectDiskData().catch((e) => debug("disk", e)),
-    collectNetworkData().catch((e) => debug("network", e)),
-  ]);
-
-  miscNodeBootTsSeconds = getSystemBootTime();
-}
+class System {
```

Comment:
As this class is designed in a way to only be used as a singleton, so try adding a constructor and throw an error if more instances are created.

Reply:
@nazarhussain I am still trying to wrap my head around what pattern is best to use here, but I still think that what I am currently doing is correct for reasons as explained in the [comment above](https://github.com/ChainSafe/lodestar/pull/5202#discussion_r1118639955).


Just to get a better picture of what you are suggesting, would the diff look something like this?

```diff
diff --git a/packages/beacon-node/src/monitoring/system.ts b/packages/beacon-node/src/monitoring/system.ts
index a13ba28db6..f69adc8c94 100644
--- a/packages/beacon-node/src/monitoring/system.ts
+++ b/packages/beacon-node/src/monitoring/system.ts
@@ -29,6 +29,19 @@ class System {
   private _miscNodeBootTsSeconds = 0;
   private _miscOs: MiscOs = "unk";
 
+  private static instance?: System;
+
+  static getInstance(): System {
+    return System.instance ?? (System.instance = new System());
+  }
+
+  private constructor() {
+    if (System.instance) {
+      throw new Error("Use System.getInstance instead of new");
+    }
+    System.instance = this;
+  }
+
   /**
    * Collect system data and update cached values
    */
@@ -212,4 +225,4 @@ class System {
   }
 }
 
-export default new System();
+export default System.getInstance();

```

Reply:
The above structure looks good to me. You can optimize it further with `static getter` and a `returning construtor`

```ts
let num = 0; 

class System {
    private static _instance: System; 
    public prop = 120; 

    constructor() {
        if(System._instance) return System._instance; 
        System._instance = this; 
        this.prop = num++; 
    }

    static get instance(): System {
        if(System._instance) return this._instance;

        this._instance = new System();
        return this._instance;
    }
}

const sys1 = System.instance; 
const sys2 = new System(); 
console.log(sys1.prop, sys2.prop);

```

This way no matter if someone call `System.instance` or `new System()` there will always be **ONLY ONE** instance.